pkg/roachprod/vm/gce/gcloud.go

```diff
@@ -1357,8 +1555,123 @@ func propagateDiskLabels(
 	return g.Wait()
 }
 
-// Delete TODO(peter): document
+type jsonManagedInstanceGroup struct {
+	Name string `json:"name"`
+	Zone string `json:"zone"`
+}
+
+// listManagedInstanceGroups returns a list of managed instance groups for a
+// given group name. Groups may exist in multiple zones with the same name. This
+// function returns a list of all groups with the given name.
+func listManagedInstanceGroups(project, groupName string) ([]jsonManagedInstanceGroup, error) {
+	args := []string{"compute", "instance-groups", "list", "--only-managed",
+		"--project", project, "--format", "json", "--filter", fmt.Sprintf("name=%s", groupName)}
+	var groups []jsonManagedInstanceGroup
+	if err := runJSONCommand(args, &groups); err != nil {
+		return nil, err
+	}
+	return groups, nil
+}
+
+// deleteInstanceTemplate deletes the instance template for the cluster.
+func deleteInstanceTemplate(project string, clusterName string) error {
+	templateName := instanceTemplateName(clusterName)
+	args := []string{"compute", "instance-templates", "delete", "--project", project, "--quiet", templateName}
+	cmd := exec.Command("gcloud", args...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "Command: gcloud %s\nOutput: %s", args, output)
+	}
+	return nil
+}
+
+// isManaged returns true if the cluster is part of a managed instance group.
+// This function makes the assumption that a cluster is either completely
+// managed or not at all.
+func isManaged(vms vm.List) bool {
+	firstVM := vms[0]
+	for k, v := range firstVM.Labels {
+		if k == ManagedLabel && v == "true" {
+			return true
+		}
+	}
+	return false
+}
+
+// Delete is part of the vm.Provider interface.
 func (p *Provider) Delete(l *logger.Logger, vms vm.List) error {
+	switch {
+	case isManaged(vms):
+		return p.deleteManaged(l, vms)
+	default:
+		return p.deleteUnmanaged(l, vms)
+	}
+}
+
+// deleteManaged deletes the managed instance group for the given VMs. It also
+// deletes any instance templates that were used to create the managed instance
+// group.
+func (p *Provider) deleteManaged(l *logger.Logger, vms vm.List) error {
+	projectClusterMap := make(map[string]map[string]struct{})
+	for _, v := range vms {
+		if v.Provider != ProviderName {
```

Comment:
IMO, this is a little too defensive. The other cloud-specific functions seem to assume that it's already part of the contract that the `vms` passed are from this provider.